sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1419,6 +1470,75 @@ pi_result _pi_queue::executeCommandList(pi_command_list_ptr_t CommandList,
   return PI_SUCCESS;
 }
 
+pi_result _pi_queue::executeImmCommandList(pi_command_list_ptr_t CommandList,
+                                           bool IsBlocking) {
+  // If the current LastCommandEvent is the nullptr, then it means
+  // either that no command has ever been issued to the queue
+  // or it means that the LastCommandEvent has been signalled and
+  // therefore that this Queue is idle.
+  // The list can be empty if command-list only contains signals of proxy
+  // events.
+  if (!CommandList->second.EventList.empty())
+    this->LastCommandEvent = CommandList->second.EventList.back();
+
+  auto &ZeCommandQueue = CommandList->second.ZeQueue;
+  // Scope of the lock must be till the end of the function, otherwise new mem
+  // allocs can be created between the moment when we made a snapshot and the
+  // moment when command list is closed and executed. But mutex is locked only
+  // if indirect access tracking enabled, because std::defer_lock is used.
+  // unique_lock destructor at the end of the function will unlock the mutex if
+  // it was locked (which happens only if IndirectAccessTrackingEnabled is
+  // true).
+  std::unique_lock<std::mutex> ContextsLock(Device->Platform->ContextsMutex,
+                                            std::defer_lock);
+  if (IndirectAccessTrackingEnabled) {
+    // We are going to submit kernels for execution. If indirect access flag is
+    // set for a kernel then we need to make a snapshot of existing memory
+    // allocations in all contexts in the platform. We need to lock the mutex
+    // guarding the list of contexts in the platform to prevent creation of new
+    // memory alocations in any context before we submit the kernel for
+    // execution.
+    ContextsLock.lock();
+    for (auto &Kernel : KernelsToBeSubmitted) {
+      if (!Kernel->hasIndirectAccess())
+        continue;
+
+      auto &Contexts = Device->Platform->Contexts;
+      for (auto &Ctx : Contexts) {
+        for (auto &Elem : Ctx->MemAllocs) {
+          const auto &Pair = Kernel->MemAllocs.insert(&Elem);
+          // Kernel is referencing this memory allocation from now.
+          // If this memory allocation was already captured for this kernel, it
+          // means that kernel is submitted several times. Increase reference
+          // count only once because we release all allocations only when
+          // SubmissionsCount turns to 0. We don't want to know how many times
+          // allocation was retained by each submission.
+          if (Pair.second)
+            Elem.second.RefCount++;
+        }
+      }
+      Kernel->SubmissionsCount++;
+    }
+    KernelsToBeSubmitted.clear();
+  }
```

Comment:
Isn't this code share between imm and not-imm submissions? If such, it should not be duplicated

Reply:
OK.